File: ios/YamabikoChat/Features/Fusion/FusionJudgeParser.swift
import Foundation

enum FusionJudgeParser {
    static func parse(_ raw: String) -> JudgeAnalysis? {
        let trimmed = extractJSON(from: raw)
        guard !trimmed.isEmpty else { return nil }
        guard let data = trimmed.data(using: .utf8) else { return nil }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        guard let dto = try? decoder.decode(JudgeAnalysisDTO.self, from: data) else {
            return nil
        }
        return dto.toJudgeAnalysis()
    }

    static func extractJSON(from text: String) -> String {
        var candidate = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if candidate.hasPrefix("```") {
            candidate = candidate
                .replacingOccurrences(of: "```json", with: "")
                .replacingOccurrences(of: "```", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if let start = candidate.firstIndex(of: "{"),
           let end = candidate.lastIndex(of: "}") {
            return String(candidate[start ... end])
        }
        return candidate
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/fusion/FusionJudgeParser.kt
package com.porarri.yamabikochat.data.fusion

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

object FusionJudgeParser {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
    }

    fun parse(raw: String): JudgeAnalysis? {
        val trimmed = extractJSON(from = raw)
        if (trimmed.isEmpty()) return null
        return runCatching {
            json.decodeFromString(JudgeAnalysisDto.serializer(), trimmed).toJudgeAnalysis()
        }.getOrNull()
    }

    fun extractJSON(from: String): String {
        var candidate = from.trim()
        if (candidate.startsWith("```")) {
            candidate = candidate
                .replace("```json", "")
                .replace("```", "")
                .trim()
        }
        val start = candidate.indexOf('{')
        val end = candidate.lastIndexOf('}')
        if (start >= 0 && end > start) {
            return candidate.substring(start, end + 1)
        }
        return candidate
    }

    fun encodeJudgeAnalysis(analysis: JudgeAnalysis): String? {
        return runCatching {
            json.encodeToString(JudgeAnalysis.serializer(), analysis)
        }.getOrNull()
    }
}
